keps/sig-network/20190603-EndpointSlice-API.md

```diff
@@ -0,0 +1,346 @@
+
+- Storing multiple megabytes of endpoints puts strain on multiple parts of the system due to not having a paging system and a monolithic watch/storage design. [The max number of endpoints is bounded by the K8s storage layer (etcd)][max-object-size]], which has a hard limit on the size of a single object (1.5MB by default). That means attempts to write an object larger than the limit will be rejected. Additionally, there is a similar limitation in the watch path in Kubernetes apiserver. For a K8s service, if its Endpoints object is too large, endpoint updates will not be propagated to kube-proxy(s), and thus iptables/ipvs won’t be reprogrammed.
+- [Performance degradation in large k8s deployments.][perf-degrade] Not being able to efficiently read/update individual endpoint changes can lead to (e.g during rolling upgrade of a service) endpoints operations that are quadratic in the number of its elements. If one consider watches in the picture (there's one from each kube-proxy), the situation becomes even worse as the quadratic traffic gets multiplied further with number of watches (usually equal to #nodes in the cluster).
+
+The new EndpointSlice API aims to address existing problems as well as leaving room for future extension.
+
+
+### Goal
+
+- Support tens of thousands of backend endpoints in a single service on cluster with thousands of nodes.
+- Move the API towards a general-purpose backend discovery API.
+- Leave room for foreseeable extension:
+  - Support multiple IPs per pod
+  - More endpoint states than Ready/NotReady
+  - Dynamic endpoint subsetting
+  
+### Non-Goal
+- Change functionality provided by K8s V1 Service API.
+- Provide better load balancing for K8s service backends.
+
+## Proposal
+
+### EndpointSlice API
+The following new EndpointSlice API will be added to the `Discovery` API group.
+
+```
+type EndpointSlice struct {
+        metav1.TypeMeta
+        // OwnerReferences should be set when the object is derived from a service.
+        metav1.ObjectMeta
+        Spec EndpointSliceSpec
+}
+
+type EndpointSliceSpec struct {
+        Endpoints []Endpoint
+        // Each EndpiontPort must have a unique port name
+        Ports []EndpointPort
+}
+
+type EndpointPort struct {
+        // Required: The name of this port.
+        // Must be a DNS_LABEL or an empty string.
+        Name string 
+        // Required: The IP protocol for this port.
+        // Must be UDP, TCP, or SCTP.
+        // Default is TCP.
+        Protocol v1.Protocol
+        // Optional: The port number of the endpoint.
+        // If this is not specified, ports are not restricted and must be interpreted in the context of the specific consumer.
+        Port *int32
+}
+
+type Endpoint struct {
+        // Required: must contain at least one backend.
+        // This can be an IP, hostname or URL.
+        // The consumer (e.g. kube-proxy) decides how to handle the backends.
+        Backends []string
```

Comment:
This needs some explanation around how eg kube-proxy would handle Backends of type URL/hostname. eg if a given Service Proxy can't handle one of the types what is the expected behavior?

Reply:
We currently stated that the consumer determines how it would handle different type of backend.

Reply:
What is the use case for URL or hostname?

What is the use case for more than one entry?

If more than one, how are consumers supposed to choose which to use?

What if some of the backends are ready but not all?

Reply:
Hostname already exists in Today's Endpoints object. It is used in KubeDns.

Adding more than one entry is originally aimed to support dual stack where each pod has 1 v4 and 1 v6 IP. 

I think in the Dual Stack KEP, it states that this effectively kicks it back to the consumer. For instance, kube-proxy can pick the 1st v4 IP and the 1st v6 IP to program. 

As one pod might have multiple IPs, we only consider pod readiness. So if one pod is ready, then all backends are ready.

For URL, there is no consumer yet. But the goal is evolve K8s Endpoints API towards a generic service discovery API so that it can be consumed by 3rd party solutions.

Reply:
I'd strongly consider splitting these into a more helpful system. E.g.:

```
Hostnames []string
IPAddresses []string
URLs []string
FQDNS []string
...
```

Alternatively, you could do a list with `type` and `value` fields. The way it is now, you can't really validate an IP address (maybe it's a funny looking DNS name with an in-house TLD).

As I understand it, these are different ways of reaching the exact same backend. I think "Aliases" or "Names" would be much less confusing names than "Backends".

Reply:
Yes. I am considering adding a field to describe backend type. So that all Endpoint in the EndpointSlice will have the same type of backends.